python_modules/dagster/dagster/_core/definitions/cacheable_assets.py

```diff
@@ -104,3 +111,161 @@ def build_definitions(
     ) -> Sequence[AssetsDefinition]:
         """For a given set of AssetsDefinitionMetadata, return a list of AssetsDefinitions"""
         raise NotImplementedError()
+
+    def with_resources(
+        self, resource_defs: Mapping[str, ResourceDefinition]
+    ) -> "CacheableAssetsDefinition":
+
+        return ResourceWrappedCacheableAssetsDefinition(self, resource_defs)
+
+    def with_prefix_or_group(
+        self,
+        output_asset_key_replacements: Optional[Mapping[AssetKey, AssetKey]] = None,
+        input_asset_key_replacements: Optional[Mapping[AssetKey, AssetKey]] = None,
+        group_names_by_key: Optional[Mapping[AssetKey, str]] = None,
+    ) -> "CacheableAssetsDefinition":
+
+        return PrefixOrGroupWrappedCacheableAssetsDefinition(
+            self,
+            output_asset_key_replacements=output_asset_key_replacements,
+            input_asset_key_replacements=input_asset_key_replacements,
+            group_names_by_key=group_names_by_key,
+        )
+
+
+class WrappedCacheableAssetsDefinition(CacheableAssetsDefinition):
+    """
+    Wraps an instance of CacheableAssetsDefinition, applying modify_assets to the
+    generated AssetsDefinition objects. This lets e.g. users define resources on
+    the cacheable assets at repo creation time which are not actually bound until
+    the assets themselves are created.
+    """
+
+    def __init__(
+        self,
+        unique_id: str,
+        wrapped: CacheableAssetsDefinition,
+    ):
+        super().__init__(unique_id)
+        self._wrapped = wrapped
+
+    def compute_cacheable_data(self) -> Sequence[AssetsDefinitionCacheableData]:
+        return self._wrapped.compute_cacheable_data()
+
+    def build_definitions(
+        self, data: Sequence[AssetsDefinitionCacheableData]
+    ) -> Sequence[AssetsDefinition]:
+        return [self.modify_assets(assets) for assets in self._wrapped.build_definitions(data)]
+
+    @abstractmethod
+    def modify_assets(self, assets: AssetsDefinition) -> AssetsDefinition:
```

Comment:
maybe `def transformed_assets_def(self, assets_def: AssetsDefinition)`

two reasons, first assets -> assets_def to line up w/ other places where a singular AssetsDefinition parameter is referred to as "assets_def", second reason is that technically we're not really "modifying" the assets, as we're creating new copies of them.

Reply:
`get_transformed_assets_def` also works